// windstorm/standalone/webpack.config.js
var path = require('path');
var webpack = require('webpack');
var relativeToRootPath = ".";
var CWD = process.cwd();
var f2eci = require(path.join(CWD,"./f2eci"));
var camelCase = require('camelcase');
var env = f2eci.env;
var PACKAGE_JSON = './package.json'
var package = require(path.join(CWD, PACKAGE_JSON));
var main = package.main
module.exports = {
  entry: {
    'standalone': [path.join(CWD, main)]
  },
  output: {
    path: path.join(CWD, f2eci.dist, package.version),
    publicPath: env == 'alpha' ? './' : f2eci["urlPrefix"] + package.version + '/',
    libraryTarget: "var",
    chunkFilename: '[name].[chunkhash].js',
    library: camelCase(package.name.replace('@','').replace('/','-')),
    filename: '[name].js'
  },
  externals: {
    "vue": "Vue",
    "vue-resource":"VueResource"
  }, 
  resolveLoader: {
    root: path.join(__dirname, 'node_modules'),
  },
  module: {
    loaders: [
      {
        test: /\.vue$/,
        loader: 'vue'
      },
      {
        test: /\.js$/,
        loader: 'babel',
        exclude: /node_modules/
      },
      {
        test: /\.json$/,
        loader: 'json'
      },
      {
        test: /\.html$/,
        loader: 'vue-html'
      },
      {
          test: /\.less$/,
          loader: "style!css!postcss!less"
      },
      {
          test: /\.css$/,
          loader: "style!css!postcss"
      },
      {
        test: /\.(jpe?g|png|gif|svg)$/i,
        loaders: [
          'url?limit=25000'
        ]
      }
    ]
  }
}
