// rainbow-enums/enums/draw-status-success.js
/**
 * @deprecated
  抽奖检查成功后的逻辑状态，根据后端返回的 data.continueDraw, data.share 和 data.maxDrawCounts 合并判断
  ===============================
  1.我要抽奖：活动进行中，有奖品库存，且用户此时有抽奖的资格
  continueDraw：true
  share: true
  maxDrawCounts: false
  2.分享再抽一次：活动进行中，有奖品库存，用户抽奖次数没超过上限值M，但当前没有抽奖资格。
  continueDraw：false
  share:true
  3. 明天再来：活动进行中，有奖品库存，今天的抽奖次数超过上限M值，但是未超过总次数的上限值。
  continueDraw：false
  share:false
  maxDrawCounts:false
  4.已领取，赶快使用吧：活动没结束，但用户抽奖总次数超过上限值M。
  continueDraw：false
  share:false
  maxDrawCounts:true
  =============================
  以上三个值做位运算，如 1，就是 二进制 100，则是4
 **/
export const DRAW_NORMAL = 'DRAW_NORMAL';
export const DRAW_SHARE_ENABLE = 'DRAW_SHARE_ENABLE';
export const DRAW_TOMORROW = 'DRAW_TOMORROW';
export const DRAW_LIMITED = 'DRAW_LIMITED';


/**
 * https://wiki.sankuai.com/pages/viewpage.action?pageId=968496820
 // 0** 可抽奖
 DRAWABLE("001","可抽奖"),
 // 1** 不可抽奖
 SHARE("101","可分享"),//有可中奖品，且可以通过分享获得抽奖机会
 HIT_ALL("102","已中全部奖品"),
 HIT_SOME("103","已中奖(部分)"),
 NO_TODAY("104","今日已无抽奖机会"),
 NO_LASTDAY("105","最后一日已无抽奖机会"),
 NO_ALL("106","活动已无抽奖机会"),
 NO_STOCK("107","活动没有库存啦");
 */
export const DRAW_DRAWABLE = 'DRAW_DRAWABLE';
export const DRAW_SHARE = 'DRAW_SHARE';
export const DRAW_HIT_ALL = 'DRAW_HIT_ALL';
export const DRAW_HIT_SOME = 'DRAW_HIT_SOME';
export const DRAW_NO_TODAY = 'DRAW_NO_TODAY';
export const DRAW_NO_LASTDAY = 'DRAW_NO_LASTDAY';
export const DRAW_NO_ALL = 'DRAW_NO_ALL';
export const DRAW_NO_STOCK = 'DRAW_NO_STOCK';
